# HG changeset patch
# User Petr Sumbera <petr.sumbera@oracle.com>
# Date 1648114576 25200
#      Thu Mar 24 02:36:16 2022 -0700
# Node ID 903e5f164c0ed424492eec6f388c03cbbcf99913
# Parent  20d81e68da033746bf81acbb08490f16679853da
Bug 1716707 [s390x] Software WebRender does not support big endian

This existing big-endian patch is kept for the imgFrame SKIA readback
correction.  The older RenderCompositorSWGL framebuffer-wide swizzle was
removed from this patch for Firefox 140.11.0 because the SPARC stack now fixes
SWGL byte order at the scalar texture and upload boundaries.  Keeping the stale
compositor swizzle would corrupt output that has already been normalized.

--- firefox-153.0/image/imgFrame.cpp
+++ firefox-153.0/image/imgFrame.cpp
@@ -379,6 +379,17 @@ mAborted = true;
     return NS_ERROR_OUT_OF_MEMORY;
   }
 
+#if defined(__BYTE_ORDER__) && __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+  if (aBackend == gfx::BackendType::SKIA && canUseDataSurface) {
+      // SKIA is lying about what format it returns on big endian
+      for (int ii=0; ii < mRawSurface->GetSize().Height()*mRawSurface->Stride() / 4; ++ii) {
+        uint32_t *vals = (uint32_t*)(mRawSurface->GetData());
+        uint32_t val = ((vals[ii] << 8) & 0xFF00FF00 ) | ((vals[ii] >> 8) & 0xFF00FF );
+        vals[ii] = (val << 16) | (val >> 16);
+      }
+  }
+#endif
+
   if (!canUseDataSurface) {
     // We used an offscreen surface, which is an "optimized" surface from
     // imgFrame's perspective.
